chapter9.adoc

```diff
@@ -0,0 +1,408 @@
+====
+A suggested reset value of 0 for the `RCID` field within the `mrmcfg` register
+aligns with the default operational behavior of resource controllers, which is
+to associate all capacity and resources with `RCID=0`. This setup ensures a
+seamless initial state that conforms to expected default resource allocation
+policies. The reset value for the `MCID` field, being potentially
+implementation-defined, does not directly impact core functionality and may
+vary based on specific system requirements.
+====
+
+When Smqosid is implemented, the configurations in the `srmcfg` register
+introduced by Ssqosid extension only apply to requests made with effective
+privilege modes less than M.
+
+[[MRMCFG64]]
+.Machine Resource Management Configuration (`mrmcfg`) register for MXLEN=64
+
+[wavedrom, , ]
+....
+{reg: [
+  {bits: 12, name: 'RCID'},
+  {bits:  4, name: 'WPRI'},
+  {bits: 12, name: 'MCID'},
+  {bits: 36, name: 'WPRI'},
+], config:{lanes: 2, hspace:1024}}
+....
+
+[[MRMCFG32]]
+.Machine Resource Management Configuration (`mrmcfg`) register for MXLEN=32
+
+[wavedrom, , ]
+....
+{reg: [
+  {bits: 12, name: 'RCID'},
+  {bits:  4, name: 'WPRI'},
+  {bits: 12, name: 'MCID'},
+  {bits:  4, name: 'WPRI'},
+], config:{lanes: 1, hspace:1024}}
+....
+
+The Smqosid extension enables the scenario where two or more security domains
+share a common set of capacity and bandwidth controllers. In this setup, access
+to the QRI of these controllers is restricted to RDSM, which then provides
+mediated access for these security domains to configure capacity or bandwidth
+allocation and read the corresponding monitoring counters. A common `QRID` is
+associated with requests from these domains, and the RDSM allocates sets of
+`RCID` and `MCID` for each domain's use. Supervisor domains are then restricted
+to selecting a `RCID` and `MCID` in `srmcfg` from the allocated range. Although
+the RDSM could limit access to `srmcfg` to prevent direct domain configuration,
+enabling domains to program `srmcfg` directly is preferred for performance
+reasons.
+
+To facilitate delegation of `srmcfg` to the SD, the Smqosid extension introduces
+`SRL` (supervisor-RCID-length) and `SML` (supervisor-MCID-length) fields in the
+`msdcfg` register to configure the usable range of `RCID` and `MCID`. These
+fields are WARL, with legal values from 0 to 12. Furthermore, a
+supervisor-srmcfg-mode (`SSM`) field is introduced, with legal values of 0 and
+1. The `SSM`, along with `SRL` and `SML` settings, determine the range of `RCID`
+and `MCID` values an SD can program into the `srmcfg` register. The RDSM is
+expected to configure these fields such that each SD can select from a disjoint
+range of values for `RCID` and `MCID`.
+
+.`msdcfg` register
+
+[wavedrom, , ]
+....
+{reg: [
+  {bits:  6, name: 'SDICN'},
+  {bits:  1, name: 'SSM'},
+  {bits:  1, name: 'SDEDBGALW'},
+  {bits:  4, name: 'SRL'},
+  {bits:  4, name: 'SML'},
+  {bits:  4, name: 'SQRID'},
+  {bits:  4, name: 'MQRID'},
```

Comment:
Question: is the reason to define dedicated fields for SQRID and MQRID that MQRID shall be automatically conveyed as soon as the hart switches into M-mode? In other word, MQRID is supposed to be tied to a fixed value (different from any SQRID), whereas SQRID is intended to be reprogrammed upon switching to a different SD? If so, a clarification would help.

The mttp register introduces SDIDs to differentiate SDs.  There is no definition for an _MDID_ while M-mode is active. It does not seem to be consistent to have different definitions (SQRID+MQRID vs SDID only).

Reply:
> Question: is the reason to define dedicated fields for SQRID and MQRID that MQRID shall be automatically conveyed as soon as the hart switches into M-mode? 

The QRID, RCID, and MCID are carried with each request made by the hart to the corresponding shared resource. The shared resource may be a cache, a memory, a TLB, execution resources within a core, power rails, etc. When execution is with M-mode privilege, the MQRID is conveyed with such requests.

> If so, a clarification would help.

Will add.

> The mttp register introduces SDIDs to differentiate SDs. There is no definition for an MDID while M-mode is active. It does not seem to be consistent to have different definitions (SQRID+MQRID vs SDID only).

That is intentional. The mttp register holds the root of the memory tracking table and the SDID is used to tag translations to more efficiently synchronize updates to the MTT with the hart execution. There is no MTT active for M-mode execution.

Reply:
> The QRID, RCID, and MCID are carried with each request made by the hart to the corresponding shared resource. The shared resource may be a cache, a memory, a TLB, execution resources within a core, power rails, etc. When execution is with M-mode privilege, the MQRID is conveyed with such requests.
> 
> > If so, a clarification would help.
> 
> Will add.

Thanks!

> > The mttp register introduces SDIDs to differentiate SDs. There is no definition for an MDID while M-mode is active. It does not seem to be consistent to have different definitions (SQRID+MQRID vs SDID only).
> 
> That is intentional. The mttp register holds the root of the memory tracking table and the SDID is used to tag translations to more efficiently synchronize updates to the MTT with the hart execution. There is no MTT active for M-mode execution.

SDs can be run in bare mode without active MTT....

Reply:
> SDs can be run in bare mode without active MTT....

When executing in a SD, MTT is always active and there is no way to disable it. However, the memory protection provided by MTT may be effectively disabled by setting the Mode to Bare - a Bare MTT is still an MTT. To draw an analogue, when V=1 (execution in a virtual machine), two-stage address translation is always in effect though either of the stages can be effectively disabled by setting the mode to Bare.

Reply:
Included this note:

> The `MQRID` is typically selected once at startup by the RDSM to designate the
> QRI utilized by the RDSM itself. Conversely, the `SQRID` may be updated during
> the process of switching SDs, to denote the QRI allocated to that specific SD by
> the RDSM.

Reply:
@ved-rivos : Thanks for adding the note. Looks good to me.

Regarding above ID discussion: the point I was trying to make: Smqosid introduces dedicated QRIDs for M-mode and non-M-mode in order to identify resource allocation configurations and monitor resource usage for _any_ type of CPU transaction. The introduction of MQRID is justified by the fact that for suitable CBQRI support shared resource/capacity controllers need to differentiate M-Mode transactions from those originating from any supervisor domain. Given that MQRID is typically a value that is fixed at startup (see above), the MQRID does not convey more information than the transaction privilege level itself. In other words: conveying MQIRD is equivalent to conveying M-mode privilege level.
On the other hand, SmMTT does not define an M-mode ID (as opposed to SDID) to tag M-mode transactions. The latter does not imply any need for applying MTT checks in M-mode, it only indicates that SmMTT does not consider M-mode transactions for the scope of supervisor domains. But with CBQRI, a distinction SQRID vs MQRID is introduced. Maybe there is some subtlety that I have missed so far.

Reply:
While the QRID, RCID, and MCID are intended to be carried with the request, the SDID is not an ID that is intended to be carried with requests. It is also a hart local ID used to identify the SD executing on a hart and not required to be globally unique across all SDs.